Chatbox

emanuelschuetze opened the pull request:
Added chatbox template stuff on top of @normanjaeckel 's pull request (#1646).

@normanjaeckel Please review.

normanjaeckel commented:
Some errors:
* If a user is not allowed to chat the button should be disabled.
* The JSData does not work like expected. Sometimes I have to reload the page or move around to get the usernames.

How to get the number of unread messages? Try the following:
* Set a global counter (use a .value() service for this) to 0 when the user opens the base template controller which controlls the Chat button.
* Increase this counter when a message comes in.
* Disabled the counter when the user opens the chat controller.
* Reset and enable the counter when the user closes the chat window.

normanjaeckel commented:
@ostcar Any idea about this described JSData thing?

emanuelschuetze commented:
All done:
- Fixed missing permission check.
- Fixed missing user in JS-Data by adding loadRelations for each chatmessage.
- Added unread counter.

Rebased on top of master.

normanjaeckel commented on openslides/core/static/js/core/base.js:
```diff
@@ -257,7 +279,15 @@ angular.module('OpenSlidesApp.core', [
     removeButtons: 'Anchor,SpecialChar,Subscript,Superscript,Styles,RemoveFormat,HorizontalRule'
 })
 
+
```
Remove this line.

emanuelschuetze replied:
Done

normanjaeckel commented on openslides/core/static/js/core/site.js:
```diff
@@ -752,6 +753,47 @@ angular.module('OpenSlidesApp.core.site', [
+    }
+])
+
+// counter of new (unread) chat messages
```
Proposal: Better move this service before the controller.

emanuelschuetze replied:
Done

normanjaeckel commented on openslides/core/static/js/core/site.js:
```diff
@@ -752,6 +753,47 @@ angular.module('OpenSlidesApp.core.site', [
+            // add new message id if there is really a new message which is not yet tracked
+            if ((oldVal[oldVal.length-1].id != newVal[newVal.length-1].id) &&
+                ($.inArray(newVal[newVal.length-1].id, NewChatMessages) == -1)) {
+                console.log(newVal[newVal.length-1].id);
```
Remove this line.

emanuelschuetze replied:
Done

normanjaeckel commented:
One remark and two proposals. Then ready to rebase and merge.

emanuelschuetze commented:
All done. Rebased. Ready for merge.

normanjaeckel commented:
Ok.

ostcar commented on openslides/core/static/js/core/base.js:
```diff
@@ -100,6 +101,13 @@ angular.module('OpenSlidesApp.core', [
             Projector.findAll();
 
+            // Loads all chat messages data and their user_ids
+            ChatMessage.findAll().then( function(chatmessages) {
```
Please add a permission check here. If the user is not a manager, then he gets a 403 what is not a problem, but it saves one request that is only needed for managers

emanuelschuetze replied:
I added a permission check:
 `if (operator.hasPerms("core.can_use_chat")) ...`
But I get this error if I add `operator` to `loadGlobalData`:
"Circular dependency found: loadGlobalData <- operator <- loadGlobalData"
@ostcar How can we fix this?

ostcar commented on openslides/core/static/js/core/base.js:
```diff
@@ -181,6 +189,20 @@ angular.module('OpenSlidesApp.core', [
     });
 }])
 
+.factory('ChatMessage', ['DS', function(DS) {
```
Please use our new style here

emanuelschuetze replied:
Done

ostcar commented on openslides/core/static/js/core/site.js:
```diff
@@ -752,6 +753,45 @@ angular.module('OpenSlidesApp.core.site', [
+                {message: $scope.newMessage}
+            )
+            .success(function () {
+                $scope.newMessage = '';
```
If you have a poor connection to the server, then the message in the text disappears not until the request returned a success message. In this case the user could hit the send button multiple times and the message would be send to the server more then once. We need a solution for this. For example, the server should skip multipe messages with the same content or you should disable the send button until the request returns a success or error message

emanuelschuetze replied:
Done.

ostcar commented on openslides/core/static/templates/index.html:
```diff
@@ -146,16 +179,17 @@ <h3 class="modal-title" translate>Please sign in!</h3>
+  <hr>
+  <footer>
+    <small>
+      &copy; Copyright 2011-2015 |
```
I never understood why we need the years. I am against it, because it has no use and we have to change it every year

emanuelschuetze commented:
All done. One open question.

emanuelschuetze commented:
I added my open question again (copied from outdated diff):
I added a permission check:
`if (operator.hasPerms("core.can_use_chat")) ...`
But I get this error if I add operator to loadGlobalData:
"Circular dependency found: loadGlobalData <- operator <- loadGlobalData"
@ostcar How can we fix this?

ostcar commented:
This is hard to fix, the problem is, that the operator is in the users app, but is requested from the core app. We have to solve this by moving the operator to the core app and make it independent from the user object. I have no idea how we can do this easily. We have to check the permission after the beta1.

So please rebase

emanuelschuetze commented:
Rebased.
